### DISCUSSION TITLE
[Python] Usage of __slots__ or @dataclass(slots=True) in python classes for reducing memory usage and performance improvement

### DISCUSSION THREAD

**Discussion by Martin15135215** [IfcOpenShell/IfcOpenShell#3857] [Ideas]
I saw the Pycon 2016 talk from Nina Zakharenko about Memory Management in Python and at the [21:19 min](https://youtu.be/F6u5rhUQ6dU?si=F32yJ5XK2z14BMYe&t=1279), she introduced the dunker slots attribute to define the attributes. Because with that, you can't assign at runtime new attributes to the class instance. This reduces the memory usage and improves performance.

For example here https://github.com/IfcOpenShell/IfcOpenShell/blob/7229f7271cd0b870a4fc5ea878cafcd05856de4e/src/ifcopenshell-python/ifcopenshell/api/spatial/assign_container.py#L24-L25

```
>>> class Usecase:
...     __slots__ = ("file","settings")
...     def __init__(self, file, product=None, relating_structure=None):
...             self.file = file
...             self.settings = {"product": product,"relating_structure": relating_structure,}
```

```
>>> Test = Usecase(file = "hallo")
>>> Test.settings
{'product': None, 'relating_structure': None}
>>> Test.new = "test"
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
AttributeError: 'Usecase' object has no attribute 'new'
```

To reduce boilerplate, there is also the possibility to use `@dataclass(slots=True)` as shown in [this video](https://youtu.be/Fot3_9eDmOs?si=jdafj818WAuZHMIp&t=400), see also the python documentation -> https://docs.python.org/3.10/library/dataclasses.html

**Comment by aothms** [comment]
Yeah this would make a difference mostly in cases when there are many instances. For the high level api usecases I don't think it'll make a big difference. The ifcopenshell entity instances fully live in C++ and get their attribute lookups dynamically, so also there I don't see much to gain.
